File: src/common/Config/Expander.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;


namespace Mason.Config
{
    public class Expander
    {
        private const string MatchGroupName = "exp";
        private const string RegExPattern = @"(?<"+MatchGroupName+@">(?:\${)(?:<"+MatchGroupName+@">|(?:[^}])+)+(?:}))";
        private const RegexOptions Options = RegexOptions.Compiled|RegexOptions.CultureInvariant|RegexOptions.IgnoreCase|RegexOptions.Multiline|RegexOptions.ExplicitCapture;
        private static readonly Regex _expression = new Regex(RegExPattern, Options);

        public Expander() { }

        private IList<string> GetSubstitutableKeys(string text)
        {
            var matches = _expression.Matches(text);
            var result = new List<string>( );
            foreach (Match match in matches)
            foreach (var groupName in _expression.GetGroupNames())
            {
                result.Add(match.Groups[groupName].Value);
            }
            return result;
        }

        public string Expand(IBuildConfig config, string raw)
        {
            return TryExpand(config, raw, out string expanded) ? expanded : raw;
        }

        public bool TryExpand(IBuildConfig config, string raw, out string result)
        {
            return TryExpand(config, 0, raw, out result);
        }
        private bool TryExpand(IBuildConfig config, int previousCount, string raw, out string result)
        {
            result = raw;
            var keys = GetSubstitutableKeys(raw);
            if (keys.Count == 0)
            {
                return true;
            }
            if (keys.Count != previousCount)
            {
                foreach (var key in keys)
                {
                    var nakedKey = key.Substring(2, key.Length - 3);
                    var replacement = config[nakedKey];
                    if (replacement == null)
                    {
                        return false;
                    }
                    result = result.Replace(key, config[nakedKey]);
                }
                return TryExpand(config, keys.Count, result, out result);
            }
            return false;
        }
    }
}

File: src/core/primitives/Expander.fs
namespace Mason
open System
open System.Linq
open System.Text.RegularExpressions

module Expander =

    [<Literal>] 
    let MatchGroupName: string = "exp"
    [<Literal>] 
    let RegExPattern: string = @"(?<" + MatchGroupName + @">(?:\${)(?:<" + MatchGroupName + @">|(?:[^}])+)+(?:}))";
    [<Literal>]
    let Options: RegexOptions = RegexOptions.Compiled|||RegexOptions.CultureInvariant|||RegexOptions.IgnoreCase|||RegexOptions.Multiline|||RegexOptions.ExplicitCapture;

    let _expression = new Regex(RegExPattern, Options);

    let rec recExpand(props : IMasonProperties, input: string): string =
        let mutable result = input
        for token in _expression.Matches(input).Cast<Match>().SelectMany(fun m -> _expression.GetGroupNames().Select(fun g -> m.Groups.[g].Value)).ToList() do
            let nakedToken = token.[2..token.Length - 2]
            result <- match null2opt props.[nakedToken] with | Some value -> recExpand(props, result.Replace(token, value)) | None -> result
        result

    let Expand(properties: IMasonProperties, text: string): string = 
        match null2opt properties with
        | Some props ->
            match null2opt text with
            | Some txt -> recExpand(props, txt)
            | None -> raise (ArgumentNullException("text"))
        | None -> raise (ArgumentNullException("properties"))
        

